src/Internal/Service/Level3Service.php

```diff
@@ -0,0 +1,192 @@
+			$level3_data['shipping_from_zip'] = $store_postcode;
+		}
+
+		return $level3_data;
+	}
+
+	/**
+	 * Processes a single order item.
+	 * Based on the queried items, this class should only receive
+	 * `WC_Order_Item_Product` or `WC_Order_Item_Fee` line items.
+	 *
+	 * @param WC_Order_Item_Product|WC_Order_Item_Fee $item     Item to process.
+	 * @param string                                  $currency Currency to use.
+	 * @return \stdClass
+	 */
+	private function process_item( WC_Order_Item $item, string $currency ): stdClass {
+		// Check to see if it is a WC_Order_Item_Product or a WC_Order_Item_Fee.
+		if ( $item instanceof WC_Order_Item_Product ) {
+			$subtotal     = $item->get_subtotal();
+			$product_id   = $item->get_variation_id()
+				? $item->get_variation_id()
+				: $item->get_product_id();
+			$product_code = substr( $product_id, 0, 12 );
+		} else {
+			$subtotal     = $item->get_total();
+			$product_code = substr( sanitize_title( $item->get_name() ), 0, 12 );
+		}
+
+		$description = substr( $item->get_name(), 0, 26 );
+		$quantity    = ceil( $item->get_quantity() );
+		$tax_amount  = $this->prepare_amount( $item->get_total_tax(), $currency );
+		if ( $subtotal >= 0 ) {
+			$unit_cost       = $this->prepare_amount( $subtotal / $quantity, $currency );
+			$discount_amount = $this->prepare_amount( $subtotal - $item->get_total(), $currency );
+		} else {
+			// It's possible to create products with negative price - represent it as free one with discount.
+			$discount_amount = abs( $this->prepare_amount( $subtotal / $quantity, $currency ) );
+			$unit_cost       = 0;
+		}
+
+		return (object) [
+			'product_code'        => (string) $product_code, // Up to 12 characters that uniquely identify the product.
+			'product_description' => $description, // Up to 26 characters long describing the product.
+			'unit_cost'           => $unit_cost, // Cost of the product, in cents, as a non-negative integer.
+			'quantity'            => $quantity, // The number of items of this type sold, as a non-negative integer.
+			'tax_amount'          => $tax_amount, // The amount of tax this item had added to it, in cents, as a non-negative integer.
+			'discount_amount'     => $discount_amount, // The amount an item was discounted—if there was a sale,for example, as a non-negative integer.
+		];
+	}
+
+	/**
+	 * Returns a bundle of products passed as an argument. Useful when working with Stripe's level 3 data
+	 *
+	 * @param array $items The Stripe's level 3 array of items.
+	 *
+	 * @return \stdClass A bundle of the products passed.
+	 */
+	private function bundle_level3_data_from_items( array $items ) {
+		// Total cost is the sum of each product cost * quantity.
+		$items_count = count( $items );
+		$total_cost  = array_sum(
+			array_map(
+				function( $cost, $qty ) {
+					return $cost * $qty;
+				},
+				array_column( $items, 'unit_cost' ),
+				array_column( $items, 'quantity' )
+			)
+		);
+
+		return (object) [
```

Comment:
Same idea as https://github.com/Automattic/woocommerce-payments/pull/7333/files#r1342602551

Reply:
And same comment :)